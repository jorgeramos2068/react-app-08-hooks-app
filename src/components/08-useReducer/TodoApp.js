import React, {useReducer, useEffect} from 'react';
import todoReducer from './todoReducer';
import TodoList from './TodoList';
import TodoForm from './TodoForm';
import './styles.css';

const init = () => {
  return JSON.parse(localStorage.getItem('todos')) || [];
};

const TodoApp = () => {
  const [todos, dispatch] = useReducer(todoReducer, [], init);

  useEffect(() => {
    localStorage.setItem('todos', JSON.stringify(todos));
  }, [todos]);

  const handleDelete = (todoId) => {
    if (isNaN(todoId)) {
      return;
    }
    const action = {
      type: 'delete',
      payload: todoId
    };
    dispatch(action);
  }

  const handleToggle = (todoId) => {
    if (isNaN(todoId)) {
      return;
    }
    const action = {
      type: 'toggle',
      payload: todoId
    };
    dispatch(action);
  };

  const handleAddTodo = (newTodo) => {
    const action = {
      type: 'add',
      payload: newTodo
    };
    dispatch(action);
  };

  return (
    <div>
      <h1>TodoApp ({todos.length})</h1>
      <hr />
      <div className="row">
        <div className="col-7">
          <TodoList
            todos={todos}
            handleDelete={handleDelete}
            handleToggle={handleToggle}
          />
        </div>
        <div className="col-5">
          <TodoForm
            handleAddTodo={handleAddTodo}
          />
        </div>
      </div>
    </div>
  );
};

export default TodoApp;
