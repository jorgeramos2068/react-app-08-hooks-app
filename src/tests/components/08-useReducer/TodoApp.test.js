import React from 'react';
import {shallow, mount} from 'enzyme';
import {act} from '@testing-library/react';
import TodoApp from '../../../components/08-useReducer/TodoApp';
import testTodos from '../../fixtures/testTodos';

describe('Tests in <TodoApp />', () => {
  let wrapper;
  Storage.prototype.setItem = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
    wrapper = shallow(<TodoApp />);
  });

  test('Should display <TodoApp /> correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('Should add a TODO', () => {
    const wrapper = mount(<TodoApp />);
    act(() => {
      wrapper.find('TodoForm').prop('handleAddTodo')(testTodos[0]);
      wrapper.find('TodoForm').prop('handleAddTodo')(testTodos[1]);
    });
    expect(wrapper.find('h1').text().trim()).toBe('TodoApp (2)');
    expect(localStorage.setItem).toHaveBeenCalledTimes(2);
  });

  test('Should delete a TODO', () => {
    wrapper.find('TodoForm').prop('handleAddTodo')(testTodos[0]);
    expect(wrapper.find('h1').text().trim()).toBe('TodoApp (1)');
    wrapper.find('TodoList').prop('handleDelete')(testTodos[0].id);
    expect(wrapper.find('h1').text().trim()).toBe('TodoApp (0)');
  });
});
