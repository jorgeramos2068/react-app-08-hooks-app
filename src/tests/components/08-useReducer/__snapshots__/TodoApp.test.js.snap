// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <TodoApp /> Should display <TodoApp /> correctly 1`] = `
<div>
  <h1>
    TodoApp (
    0
    )
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TodoList
        handleDelete={[Function]}
        handleToggle={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col-5"
    >
      <TodoForm
        handleAddTodo={[Function]}
      />
    </div>
  </div>
</div>
`;
